### PR TITLE
network: handle embedded nulls in unix domain socket

### PR DESCRIPTION
**Pull request by asraa**
This PR rejects embedded nulls for pathname domain socket addresses, and handles them in abstract domain socket addresses. For abstract domain socket addresses, the `friendly_name_` replaces null bytes with '@' (for consistency with the first byte) so that log messages are preserved.

See http://man7.org/linux/man-pages/man7/unix.7.html.

Fixes OSS-Fuzz bug
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=17984
(This was occurring due to null characters in the address resulting in an empty, but valid, address).
Testing: Added unit tests to test embedded nulls accepted and formatted with '@' in abstract namespace, and rejected for pathnames.

Signed-off-by: Asra Ali <asraa@google.com>

### CODE REVIEW

File: test/common/network/address_impl_test.cc [hunk 1]
```diff
@@ -336,6 +336,29 @@ TEST(PipeInstanceTest, BadAddress) {
...
+#if defined(__linux__)
+  std::string embedded_null("@/foo/bar");
+  embedded_null[5] = '\0'; // Set embedded null.
+  PipeInstance address(embedded_null);
+  EXPECT_EQ("@/foo@bar", address.asString());
+  EXPECT_EQ("@/foo@bar", address.asStringView());
+  EXPECT_EQ(Type::Pipe, address.type());
+  EXPECT_EQ(nullptr, address.ip());
+#else
+  EXPECT_THROW(PipeInstance address(embedded_null), EnvoyException);
```
Comment (zuercher) [inline test/common/network/address_impl_test.cc]: Just a drive-by: I don't think this compiles on macOS because the definition of `embedded_null` is within the `#if defined(__linux__)` above

Comment (asraa) [reply test/common/network/address_impl_test.cc]: Thank you!! Just fixed that.

File: source/common/network/address_impl.cc [hunk 2]
```diff
@@ -43,6 +43,13 @@ void validateIpv6Supported(const std::string& address) {
   }
 }
 
+// Returns the abstract path with embedded nulls replaces with '@'.
+std::string friendlyNameFromAbstractPath(const char* path, size_t num) {
```
Comment (lizan) [inline source/common/network/address_impl.cc]: use absl::string_view as parameter instead of const char* and size_t pair?

Comment (asraa) [reply source/common/network/address_impl.cc]: I chose that way only because I still need to explicitly pass the length into a string_view parameter since the abstract path has embedded nulls. 

To avoid copies I could also pass the `friendly_name_` into the function as a reference to avoid copy constructing the return value?

File: source/common/network/address_impl.cc [hunk 3]
```diff
@@ -43,6 +43,13 @@ void validateIpv6Supported(const std::string& address) {
   }
 }
 
+// Construsts a readable string with the embedded nulls in the abstract path replaced with '@'.
+std::string friendlyNameFromAbstractPath(absl::string_view path) {
+  std::string friendly_name(path.data(), path.size());
```
Comment (lizan) [inline source/common/network/address_impl.cc]: Use absl::StrReplaceAll? then you don't have to explicitly making a copy here.

Comment (asraa) [reply source/common/network/address_impl.cc]: absl::StrRepalceAll doesn't play well with replacing single null characters because it grabs the replacement by a string_view without the size parameter:
https://github.com/abseil/abseil-cpp/blob/19b021cb3ff23048dfbe236a4e611925d8930831/absl/strings/str_replace.h#L164

I think at this point even if it looks messy, I could pass in a reference to `friendly_name` and modify it in the function with `std::replace`.

Comment (lizan) [reply source/common/network/address_impl.cc]: If you make StrToStrMap typed as string_view (or std::string) instead of `const char *`, then it can play, but not a big deal here.
